File: cortex/brain/tools.py
"""Herramientas del brain — read-only + safe-action, sobre servicios existentes.

Contrato (doc 06 §BRAIN v1):
- ``Tier.READ``: consulta pura, sin side-effects.
- ``Tier.SAFE_ACTION``: side-effect externo no destructivo whitelisteado
  (único permitido hoy: webgraph.serve).
- Las MUTACIONES no son herramientas: ``actions_propose`` devuelve el
  comando CLI exacto para que el usuario las ejecute ("propone, no ejecuta").
"""

from __future__ import annotations

import subprocess
import sys
from collections.abc import Callable
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import Any

from cortex.action_engine.context import ActionContext


class Tier(str, Enum):
    READ = "read"
    SAFE_ACTION = "safe_action"


@dataclass(frozen=True)
class ToolSpec:
    name: str
    description: str
    tier: Tier
    handler: Callable[..., str]
    args_hint: str = ""


def build_tools(ctx: ActionContext) -> dict[str, ToolSpec]:
    """Construye el registro de herramientas anclado al proyecto de *ctx*."""
    tools: dict[str, ToolSpec] = {}

    def tool(name: str, description: str, tier: Tier, args_hint: str = ""):
        def deco(fn: Callable[..., str]) -> None:
            tools[name] = ToolSpec(name=name, description=description, tier=tier,
                                   handler=fn, args_hint=args_hint)

        return deco

    # ── READ ───────────────────────────────────────────────────────────

    @tool("memory.search", "Búsqueda híbrida (RRF) en memoria episódica+semántica.",
          Tier.READ, "<query> [top_k]")
    def memory_search(query: str, top_k: int = 5) -> str:
        result = ctx.mem.retrieve(query, top_k=top_k)
        if not result.unified_hits:
            return "Sin resultados en este repo para esa consulta."
        lineas = ["Resultados (RRF híbrido):"]
        for i, hit in enumerate(result.unified_hits, 1):
            lineas.append(
                f"  {i}. [{hit.source}] {hit.display_title} "
                f"({hit.display_path}) score={hit.score:.4f}"
            )
        return "\n".join(lineas)

    @tool("docs.related", "Documentos del vault relacionados con un tema "
          "(embeddings OPT-IN: precise=e5-large ~2GB RAM / fast=MiniLM liviano).",
          Tier.READ, "<tema> [precise|fast]")
    def docs_related(tema: str, engine: str = "") -> str:
        from cortex.semantic.vault_reader import VaultReader

        if not engine:
            return (
                "¿Qué precisión preferís?\n"
                "  · precise → e5-large multilingüe, máxima calidad (~2GB RAM)\n"
                "  · fast    → MiniLM, liviano y veloz (puede desviar)\n"
                "Respondé 'docs.related <tema> precise' o '... fast'."
            )
        model, backend = (
            ("intfloat/multilingual-e5-large", "fastembed")
            if engine.startswith("precise")
            else ("all-MiniLM-L6-v2", "onnx")
        )
        reader = VaultReader(
            vault_path=str(ctx.vault_path),
            embedding_model=model,
            embedding_backend=backend,
            vector_cache=None,
        )
        reader.sync()
        hits = reader.search(tema, top_k=5)
        if not hits:
            return f"Sin documentos relacionados con '{tema}' en este repo."
        lineas = [f"Documentos relacionados con '{tema}' ({model}):"]
        for i, h in enumerate(hits, 1):
            lineas.append(f"  {i}. {h.title} ({h.path})")
        return "\n".join(lineas)

    @tool("cortex.health", "Estado de salud de Cortex en este proyecto.", Tier.READ)
    def cortex_health() -> str:
        from cortex.tui.core import snapshot_home

        state = snapshot_home(ctx.layout.repo_root)
        items = "; ".join(f"{k}: {v}" for k, v in state.doctor_items)
        sesion = state.sesion_line or "ninguna activa"
        return (
            f"Salud Cortex — {items} · vault: {state.vault_notas} notas · "
            f"sesión: {sesion} · snapshot {state.elapsed_ms}ms"
        )

    @tool("vault.stats", "Conteos del vault y workspace.", Tier.READ)
    def vault_stats() -> str:
        vault = ctx.vault_path
        notas = len(list(vault.rglob("*.md"))) if vault.exists() else 0
        specs = len(list((vault / "specs").glob("*.md"))) if (vault / "specs").exists() else 0
        return f"Vault: {notas} notas .md ({specs} specs). Workspace: {ctx.dot_cortex}"

    @tool("session.current", "Sesión activa y sus checkpoints.", Tier.READ)
    def session_current() -> str:
        abiertas = [r for r in ctx.sessions.list() if r.status.value == "open"]
        if not abiertas:
            return "No hay sesión OPEN. Abrí una con `cortex start`."
        r = abiertas[0]
        return (
            f"Sesión {r.session_id} · OPEN · {len(r.checkpoints)} checkpoints · "
            f"abierta {r.opened_at:%Y-%m-%d %H:%M}"
        )

    # ── SAFE_ACTION ────────────────────────────────────────────────────

    @tool("webgraph.serve",
          "Levanta el visualizador del webgraph y reporta el puerto.",
          Tier.SAFE_ACTION)
    def webgraph_serve() -> str:
        from cortex.webgraph.config import WebGraphConfig

        config = WebGraphConfig.load(ctx.layout.workspace_root.parent if False else Path.cwd(),
                                     )
        puerto = config.server_port or 8000
        subprocess.Popen(
            [sys.executable, "-m", "cortex.cli.main", "webgraph", "serve", "--no-open"],
            cwd=str(Path.cwd()),
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        return f"Webgraph abierto en http://127.0.0.1:{puerto} — mirá ese puerto."

    # ── propose-only: mutaciones NUNCA ejecutables acá ────────────────

    @tool("actions.propose",
          "Lista acciones sugeridas por el ActionEngine CON el comando exacto "
          "para ejecutarlas vos (el brain nunca muta).",
          Tier.READ)
    def actions_propose() -> str:
        from cortex.action_engine.actions import build_default_registry
        from cortex.action_engine.scheduler import Scheduler
        from cortex.action_engine.signals import leer_senales
        from cortex.action_engine.store import PreferencesStore

        registry = build_default_registry(ctx)
        sched = Scheduler(
            preferences=PreferencesStore(ctx.dot_cortex),
            senales=leer_senales(ctx.dot_cortex),
        )
        propuestas = sched.propose(registry)
        if not propuestas:
            return "Nada pendiente ✓"
        lineas = ["Acciones sugeridas (ejecutalas VOS con el comando indicado):"]
        for p in propuestas:
            lineas.append(f"  · {p.action.id} — {p.action.title}")
            lineas.append(f"      → cortex next --json   |   efecto: {p.action.effect}")
        lineas.append("El brain propone; la ejecución es tuya (modo estricto).")
        return "\n".join(lineas)

    return tools

File: rust/crates/cortex-brain/src/tools.rs
//! Herramientas del brain nativo — READ + SAFE_ACTION sobre el CLI `cortex`.
//!
//! Contrato (doc 06 §BRAIN v1, decisión dueño 2026-08-24: BRAIN-2/3 nativos):
//! - `Tier::Read`: consulta pura, sin side-effects.
//! - `Tier::SafeAction`: único permitido hoy: webgraph.serve.
//! - Las MUTACIONES no son tools: actions.propose devuelve el comando CLI
//!   exacto para que el usuario lo ejecute ("propone, no ejecuta").
//!
//! Los servicios session/actions/health siguen siendo Python hasta Obra E;
//! las tools los consumen INVOCANDO EL CLI `cortex` como cualquier usuario —
//! contrato estable, cero duplicación. Override de binario: env CORTEX_BIN.

use std::collections::BTreeMap;
use std::process::Command;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Tier {
    Read,
    SafeAction,
}

#[derive(Debug, Clone)]
pub struct ToolSpec {
    pub name: &'static str,
    pub description: &'static str,
    pub tier: Tier,
    pub args_hint: &'static str,
}

/// Registro ordenado por nombre (determinista para /help y tests).
pub fn build_tools() -> BTreeMap<&'static str, ToolSpec> {
    let mut t = BTreeMap::new();
    for (name, desc, tier, hint) in [
        ("memory.search", "Búsqueda híbrida (RRF) en memoria episódica+semántica.", Tier::Read, "<query> [top_k]"),
        ("docs.related", "Documentos del vault relacionados con un tema (embeddings OPT-IN).", Tier::Read, "<tema> [precise|fast]"),
        ("cortex.health", "Estado de salud de Cortex en este proyecto.", Tier::Read, ""),
        ("vault.stats", "Conteos del vault y workspace.", Tier::Read, ""),
        ("session.current", "Sesión activa y sus checkpoints.", Tier::Read, ""),
        ("webgraph.serve", "Levanta el visualizador del webgraph y reporta el puerto.", Tier::SafeAction, ""),
        ("actions.propose", "Lista acciones sugeridas CON el comando exacto para ejecutarlas vos (el brain nunca muta).", Tier::Read, ""),
    ] {
        t.insert(name, ToolSpec { name, description: desc, tier, args_hint: hint });
    }
    t
}

/// Binario CLI a invocar (override para tests).
pub fn cortex_bin() -> String {
    std::env::var("CORTEX_BIN").unwrap_or_else(|_| "cortex".into())
}

/// Ejecuta el CLI cortex capturando stdout. Falla con mensaje accionable.
fn run_cli(args: &[&str]) -> Result<String, String> {
    let out = Command::new(cortex_bin())
        .args(args)
        .output()
        .map_err(|e| format!("no pude ejecutar {}: {e}", cortex_bin()))?;
    if !out.status.success() {
        return Err(format!(
            "{} {} falló (rc={:?}): {}",
            cortex_bin(),
            args.join(" "),
            out.status.code(),
            String::from_utf8_lossy(&out.stderr).trim()
        ));
    }
    Ok(String::from_utf8_lossy(&out.stdout).trim_end().to_string())
}

/// Ejecuta una tool por nombre con argumentos libres. Devuelve texto usuario.
///
/// NUNCA ejecuta mutaciones: las únicas herramientas con side-effect son
/// SAFE_ACTION whitelisteada (webgraph.serve, spawn detached).
pub fn dispatch(tool: &str, args: &[String]) -> Result<String, String> {
    match tool {
        "cortex.health" => run_cli(&["doctor"]),
        "session.current" => run_cli(&["context"]),
        "memory.search" | "docs.related" => {
            if args.is_empty() || args[0].trim().is_empty() {
                if tool == "docs.related" {
                    return Ok(
                        "¿Qué precisión preferís?\n  · precise → e5-large multilingüe, máxima calidad (~2GB RAM)\n  · fast    → MiniLM, liviano y veloz\nRespondé 'docs.related <tema> fast'.".into(),
                    );
                }
                return Err("falta <query>".into());
            }
            run_cli(&["search", &args.join(" ")])
        }
        "vault.stats" => {
            // Conteo nativo: vault/**/*.md (convención Cortex).
            let count = count_markdown("vault");
            Ok(format!("Vault: {count} notas .md"))
        }
        "webgraph.serve" => {
            spawn_detached(&["webgraph", "serve", "--no-open"])?;
            Ok("Webgraph abierto en http://127.0.0.1:8000 — mirá ese puerto.".into())
        }
        "actions.propose" => propose(),
        other => Err(format!("tool desconocida: {other}")),
    }
}

/// actions.propose: lista sugerencias + comando exacto. El brain NUNCA muta.
fn propose() -> Result<String, String> {
    let raw = run_cli(&["next", "--json"])?;
    let value: serde_json::Value =
        serde_json::from_str(&raw).map_err(|e| format!("next --json no es JSON válido: {e}"))?;
    let items = value.as_array().cloned().unwrap_or_default();
    if items.is_empty() {
        return Ok("Nada pendiente ✓".into());
    }
    let mut out = String::from("Acciones sugeridas (ejecutalas VOS con el comando indicado):\n");
    for item in &items {
        let id = item["id"].as_str().unwrap_or("?");
        let title = item["title"].as_str().unwrap_or("");
        out.push_str(&format!("  · {id} — {title}\n"));
        out.push_str("      → cortex next --json   |   efecto: ver doctor\n");
    }
    out.push_str("El brain propone; la ejecución es tuya (modo estricto).");
    Ok(out)
}

/// Spawn detached (nuevo session): el brain no queda atado al servidor.
fn spawn_detached(args: &[&str]) -> Result<(), String> {
    use std::process::Stdio;
    Command::new(cortex_bin())
        .args(args)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .stdin(Stdio::null())
        .spawn()
        .map(|_| ())
        .map_err(|e| format!("spawn falló: {e}"))
}

fn count_markdown(dir: &str) -> usize {
    fn walk(path: &std::path::Path, acc: &mut usize) {
        if let Ok(rd) = std::fs::read_dir(path) {
            for entry in rd.flatten() {
                let p = entry.path();
                if p.is_dir() {
                    walk(&p, acc);
                } else if p.extension().is_some_and(|e| e == "md") {
                    *acc += 1;
                }
            }
        }
    }
    let mut n = 0;
    walk(std::path::Path::new(dir), &mut n);
    n
}

#[cfg(test)]
mod tests {
    use std::sync::Mutex;
    pub(crate) static ENV_LOCK: Mutex<()> = Mutex::new(());
    use super::*;

    #[test]
    fn no_hay_herramientas_mutadoras() {
        let mutadoras = [
            "vault.reindex",
            "session.checkpoint_now",
            "setup.finish_bootstrap",
        ];
        let tools = build_tools();
        for m in mutadoras {
            assert!(!tools.contains_key(m), "{m} jamás puede ser tool");
        }
    }

    #[test]
    fn todas_read_o_safe_action() {
        for spec in build_tools().values() {
            assert!(matches!(spec.tier, Tier::Read | Tier::SafeAction));
        }
    }

    #[test]
    fn webgraph_es_safe_action() {
        assert_eq!(build_tools()["webgraph.serve"].tier, Tier::SafeAction);
    }

    #[test]
    fn tool_desconocida_falla_ruidosa() {
        assert!(dispatch("vault.reindex", &[]).is_err(), "mutación jamás");
    }

    #[test]
    fn propose_nunca_ejecuta_y_ofrece_comando() {
        let _env = ENV_LOCK.lock().unwrap();
        // CORTEX_BIN=/bin/echo: next --json devuelve texto plano → JSON inválido
        // ⇒ error ruidoso (nunca ejecuta nada). Con salida vacía → "Nada pendiente".
        unsafe { std::env::set_var("CORTEX_BIN", "/bin/echo") };
        let out = propose().unwrap_or_else(|e| e);
        // /bin/echo imprime sus args: "next --json" ≠ JSON válido ⇒ error limpio.
        assert!(out.contains("propone") || out.contains("JSON") || out.contains("Nada pendiente"));
        unsafe { std::env::remove_var("CORTEX_BIN") };
    }
}
